lib/HLSL/DxilNoOptLegalize.cpp

```diff
@@ -120,10 +124,27 @@ class DxilNoOptSimplifyInstructions : public ModulePass {
               I->eraseFromParent();
               Changed = true;
             }
+          } else if (PHINode* Phi = dyn_cast<PHINode>(I)) {
+            if (Phi->getNumIncomingValues() == 1
+             && Phi->getType() == HandleType) {
+                Changed = true;
+                LcssaHandlePhis.push_back(Phi);
+            }
           }
         }
       }
     }
+
+    // Replace all single value phi nodes (these are inserted by lcssa) of
+    // resource handles with the resource handle itself. This avoids
+    // validation errors since we do not allow handles in phis.
+    for (PHINode* Phi : LcssaHandlePhis) {
+      assert(Phi->getNumIncomingValues() == 1);
```

Comment:
We use DXASSERT in other parts of this directory. We should probably be consistent here